components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/AccessTokenIssuer.java

```diff
@@ -1394,6 +1402,91 @@ private void triggerPostListeners(OAuth2AccessTokenReqDTO tokenReqDTO,
         }
     }
 
+    private static void triggerPostIssueTokenEvent(OAuth2AccessTokenReqDTO tokenReqDTO,
+                                                   OAuth2AccessTokenRespDTO tokenRespDTO,
+                                                   OAuthTokenReqMessageContext tokReqMsgCtx)
+            throws IdentityOAuth2Exception, OrganizationManagementException {
+
+        if (!Boolean.parseBoolean(IdentityUtil.getProperty(ENABLE_POST_TOKEN_ISSUE_EVENT))) {
+            if (log.isDebugEnabled()) {
+                log.debug("Token event publishing is disabled. Hence skipping the post issue token event.");
+            }
+            return;
+        }
+        if (tokenReqDTO == null || tokenRespDTO == null || tokReqMsgCtx == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Token request DTO, token response DTO or token request message context is null. " +
+                        "Skipping the post issue token event.");
+            }
+            return;
+        }
+        if (tokenRespDTO.isError()) {
+            if (log.isDebugEnabled()) {
+                log.debug("Token response DTO is in error state. Hence skipping the post issue token event.");
+            }
+            return;
+        }
+        if (existingTokenUsed(tokReqMsgCtx)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Existing token is used. Hence skipping the post issue token event.");
+            }
+            return;
+        }
+        String userType = StringUtils.EMPTY;
+        String organizationId = StringUtils.EMPTY;
+        String tenantDomain = StringUtils.EMPTY;
+        String clientId = StringUtils.EMPTY;
+        String accessingOrganizationId = StringUtils.EMPTY;
+        String tokenId = StringUtils.EMPTY;
+        String grantType = StringUtils.EMPTY;
+        String issuedTime = String.valueOf(tokReqMsgCtx.getAccessTokenIssuedTime());
+
+        Object userTypeObject = tokReqMsgCtx.getProperty(OAuthConstants.UserType.USER_TYPE);
+        if (userTypeObject instanceof String) {
+            userType = (String) userTypeObject;
+        }
+        if (tokReqMsgCtx.getOauth2AccessTokenReqDTO() != null &&
+                tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain() != null) {
+            organizationId = OAuthComponentServiceHolder.getInstance().getOrganizationManager()
+                    .resolveOrganizationId(tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain());
+        }
+        if (tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain() != null) {
+            tenantDomain = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getTenantDomain();
+        }
+        if (tokReqMsgCtx.getOauth2AccessTokenReqDTO().getClientId() != null) {
+            clientId = tokReqMsgCtx.getOauth2AccessTokenReqDTO().getClientId();
```

Comment:
Do we need to handle null checks for userType, tenantDomain, clientId, tokenId and grantType. Because these are mandatory items in a token issuance flow.

Also we can simplify these resolving logic for b2b cases
if (StringUtils.isNotEmpty(tenantDomain)) {
    organizationId = OAuthComponentServiceHolder.getInstance()
            .getOrganizationManager()
            .resolveOrganizationId(tenantDomain);
}

String accessingOrganizationId = user != null ? user.getAccessingOrganization() : StringUtils.EMPTY;